Src/Base/AMReX_SIMD.H

```diff
@@ -0,0 +1,141 @@
+    using SIMDParticleReal = std::experimental::fixed_size_simd<amrex::ParticleReal, SIMD_WIDTH>;
+
+    // Type that has the same amount of IdCpu SIMD elements as the SIMDParticleReal type
+    template<typename T_ParticleReal = SIMDParticleReal<>>
+    using SIMDIdCpu = std::experimental::rebind_simd_t<uint64_t, T_ParticleReal>;
+#else
+    constexpr auto native_simd_size_real = 1;
+    constexpr auto native_simd_size_particlereal = 1;
+
+    template<int SIMD_WIDTH = native_simd_size_real>
+    using SIMDReal = amrex::Real;
+
+    template<int SIMD_WIDTH = native_simd_size_particlereal>
+    using SIMDParticleReal = amrex::ParticleReal;
+
+    // Type that has the same amount of IdCpu SIMD elements as the SIMDParticleReal type
+    template<typename T_ParticleReal = SIMDParticleReal<>>
+    using SIMDIdCpu = uint64_t;
+#endif
+
+    namespace detail {
+        struct InternalVectorized {};
+    }
+
+    /** Mixin Helper Class
+     *
+     * Use this class as a mixin (base) class to simplify writing functors that support/do not support
+     * ParallelForSIMD.
+     *
+     * Example:
+     * ```c++
+     * struct Compute : public Vectorized<>
+     * {
+     *     template<typename T_Real>
+     *     AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+     *     void operator() (T_Real & AMREX_RESTRICT x) { ... }
+     * };
+     *
+     * // ... call site below
+     * {
+     *     Compute c;
+     *
+     *     if constexpr (amrex::simd::is_vectorized<Compute>) {
+     *         ParallelForSIMD<...>(np, c).
+     *     } else {
+     *         ParallelFor(np, c);
+     *     }
+     * }
+     * ```
+     */
+    template<int SIMD_WIDTH = native_simd_size_real>
+    struct
+    Vectorized : detail::InternalVectorized
+    {
+        //! SIMD width in elements
+        static constexpr int simd_width = SIMD_WIDTH;
+    };
+
+    /** Check if a Functor Class works with amrex::ParallelForSIMD
+     *
+     * @see amrex::simd::Vectorized
+     */
+    template<typename T>
+    constexpr bool is_vectorized = std::is_base_of_v<detail::InternalVectorized, T>;
+
+    /** Check if a function argument is declared as non-const
+     *
+     * Use in conjunction with conditional write-back logic from vector registers, e.g.,
+     *
+     * ```c++
+     * template<typename T_Real>
+     * void compute (T_Real & AMREX_RESTRICT x,
+     *               T_Real const & AMREX_RESTRICT y) { ... }
+     *
+     * part_x.copy_from(&m_part_x[i], stdx::element_aligned);
+     * part_y.copy_from(&m_part_y[i], stdx::element_aligned);
+     *
+     * compute(part_x, part_y);
+     *
+     * if constexpr (is_nth_arg_non_const(compute<double>, 0))
+     *     part_x.copy_to(&m_part_x[i], stdx::element_aligned);
+     * if constexpr (is_nth_arg_non_const(compute<double>, 1))
+     *     part_y.copy_to(&m_part_y[i], stdx::element_aligned);
+     */
+    template<typename R, typename... Args>
+    constexpr bool is_nth_arg_non_const (R(*)(Args...), int n)
```

Comment:
Need to add support for functors as well (used as such in ImpactX).

Note: Functor `operator()` signature types always come as two types, const and non-const member functions.